lebedigital/demonstrator_scripts/kpi_from_fem.py

```diff
@@ -0,0 +1,87 @@
+from lebedigital.unit_registry import ureg
+import numpy as np
+import pint_pandas
+import pint
+import pandas as pd
+
+
+def kpi_from_fem(df,limit_temp):
+    """
+    compute KPIs from simulation output
+
+    the maximum reached temperature and the time is extracted and the difference to the limit temperature given
+    the time when the yield value equals zero is approximated
+
+    Parameters
+    ----------
+    df : panda-pint dataframe
+        data frame with the columns "time","temperature","yield"
+    limit_temp: float in pint units
+        maximum allowed temperature
+
+    Returns
+    -------
+    results : dict with pint unit
+        - 'max_reached_temperature' in degree_Celsius
+        - 'time_max_reached_temperature' in hours
+        - 'check_reached_temperature' in degree_Celsius
+        - 'time_of_demolding' in hours
+    """
+    # initialze dictionary
+    results = {}
+
+    # Part 1 : get max temp, with time -> compute difference to limit
+
+    results['max_reached_temperature'] = df['temperature'].max()
+    results['time_max_reached_temperature'] = \
+        df.loc[df['temperature'] == results['max_reached_temperature']]['time'].tolist()[0]
+
+    # changing units, because we can
+    results['time_max_reached_temperature'].ito('hours')
+
+    # difference between limit temp and reached maximum
+    results['check_reached_temperature'] = limit_temp - results['max_reached_temperature']
+
+    # Part 2 : interpolate time when yield == 0
+
+    # make sure the columns have expected units
+    df['time'] = df['time'].pint.to("seconds")
+    df['temperature'] = df['temperature'].pint.to("degree_Celsius")
+    df['yield'] = df['yield'].pint.to("dimensionless")
+
+    # subsequent operations have problems with the units (interpolation) therefore convert to standard df
+    df = df.pint.dequantify()
+
+    # adding new row with zero yield
+    new_line = pd.DataFrame({('time', 'second'): [np.nan], ('temperature', 'degree_Celsius'): [np.nan], ('yield', 'dimensionless'): [0.0]})
+    df = pd.concat([df, new_line],  ignore_index=True)
```

Comment:
If you are sorting according to the yield, what is happening if you have two instances in time where the yield surface changes the sign? In that case you might interpolate between values of different time instances? If you remove all values after the first occurence of a neg. yield value, this would not happen, but in that case you just have to interpolate between this value and the previous one?

Reply:
This case is not covered. I have not seen a simulation where that happens, as the load is constant and the material get stiffer over time.
I will add a comment to indicate that this might have to be added. but to catch this in a useful way, I would have to change this approach.